subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestClassProcessor.java

```diff
@@ -107,7 +110,37 @@ public void execute(@Nonnull String testClassName) {
         private void processAllTestClasses() {
             LauncherDiscoveryRequest discoveryRequest = createLauncherDiscoveryRequest(testClasses);
             TestExecutionListener executionListener = new JUnitPlatformTestExecutionListener(resultProcessor, clock, idGenerator);
-            launcherSession.getLauncher().execute(discoveryRequest, executionListener);
+            Launcher launcher = launcherSession.getLauncher();
+            if (spec.isDryRun()) {
+                TestPlan testPlan = launcher.discover(discoveryRequest);
+                executeDryRun(testPlan, executionListener);
+            } else {
+                launcher.execute(discoveryRequest, executionListener);
+            }
+        }
+    }
+
+    private void executeDryRun(TestPlan testPlan, TestExecutionListener listener) {
+        listener.testPlanExecutionStarted(testPlan);
+
+        for (TestIdentifier root : testPlan.getRoots()) {
+            dryRun(root, testPlan, listener);
+        }
+
+        listener.testPlanExecutionFinished(testPlan);
+    }
+
+    private void dryRun(TestIdentifier testIdentifier, TestPlan testPlan, TestExecutionListener listener) {
+        listener.executionStarted(testIdentifier);
+
+        for (TestIdentifier child : testPlan.getChildren(testIdentifier)) {
+            dryRun(child, testPlan, listener);
+        }
+
+        if (testIdentifier.isContainer()) {
```

Comment:
Since JUnit has `TestType.CONTAINER_AND_TEST` (which is used by Spock IIRC) I think we should invert the if and call `executionSkipped` when `testIdentifier.isTest()` returns `true`.